ReactiveCocoa/Swift/Flatten.swift

```diff
@@ -241,23 +241,48 @@ private final class ConcatState<Value, Error: ErrorType> {
 
 	/// Subscribes to the given signal producer.
 	func startNextSignalProducer(signalProducer: SignalProducer<Value, Error>) {
-		signalProducer.startWithSignal { signal, disposable in
-			let handle = self.disposable?.addDisposable(disposable) ?? nil
-
-			signal.observe { event in
-				switch event {
-				case .Completed, .Interrupted:
-					handle?.remove()
+		// Current signal producer
+		var signalProducer: SignalProducer<Value, Error>? = signalProducer
+		// Are we still inside the loop
+		var inLoop = true
```

Comment:
@andersio thanks for catching the problem, yeah, also the loop flag may have some corner case race condition

like, say `signalProducer = nextSignalProducer` kicks in right after `startNextSignalProducer` leaves while loop but before setting inLoop to false

```
Thread 1                                Thread 2
left while loop
                                        signalProducer = nextSignalProducer
inLoop = false  
startNextSignalProducer returns
```

then the next signal producer will never be started

I am thinking how to avoid that problem. Had an idea already